libkbfs/delayed_cancellation.go

```diff
@@ -0,0 +1,224 @@
+	CtxCancellationDelayerKey = "libkbfs-cancellation-delayer"
+)
+
+// CtxReplayFunc is a function for replaying a series of changes done on a
+// context.
+type CtxReplayFunc func(ctx context.Context) context.Context
+
+// CtxNotReplayableError is returned when NewContextWithReplayFrom is called on
+// a ctx with no replay func.
+type CtxNotReplayableError struct{}
+
+func (e CtxNotReplayableError) Error() string {
+	return "Unable to replay on ctx"
+}
+
+// NoCancellationDelayerError is returned when EnableDelayedCancellationWithGracePeriod or
+// ExitCritical are called on a ctx without Critical Awareness
+type NoCancellationDelayerError struct{}
+
+func (e NoCancellationDelayerError) Error() string {
+	return "Context doesn't have critical awareness or CtxCancellationDelayerKey " +
+		"already exists in ctx but is not of type *cancellationDelayer"
+}
+
+// ContextAlreadyHasCancellationDelayerError is returned when
+// NewContextWithCancellationDelayer is called second time on the same ctx, which
+// is not supported yet.
+type ContextAlreadyHasCancellationDelayerError struct{}
+
+func (e ContextAlreadyHasCancellationDelayerError) Error() string {
+	return "Context already has critical awareness; only one layer is supported."
+}
+
+// NewContextReplayable creates a new context from ctx, with change applied. It
+// also makes this change replayable by NewContextWithReplayFrom. When
+// replayed, the resulting context is replayable as well.
+//
+// It is important that all WithValue-isu mutations on ctx is done "replayably"
+// (with NewContextReplayable) if any delayed cancellation is used, e.g.
+// through EnableDelayedCancellationWithGracePeriod,
+func NewContextReplayable(
+	ctx context.Context, change CtxReplayFunc) context.Context {
+	var replay CtxReplayFunc
+	replay = func(ctx context.Context) context.Context {
+		ctx = change(ctx)
+		// _ is necessary to avoid panic if nil
+		replays, _ := ctx.Value(CtxReplayKey).([]CtxReplayFunc)
+		replays = append(replays, replay)
+		return context.WithValue(ctx, CtxReplayKey, replays)
+	}
+	return replay(ctx)
+}
+
+// NewContextWithReplayFrom constructs a new context out of ctx by calling all
+// attached replay functions. This disconnects any existing context.CancelFunc.
+func NewContextWithReplayFrom(ctx context.Context) (context.Context, error) {
+	if replays, ok := ctx.Value(CtxReplayKey).([]CtxReplayFunc); ok {
+		newCtx := context.Background()
+		for _, replay := range replays {
+			newCtx = replay(newCtx)
+		}
+		return newCtx, nil
+	}
+	return nil, CtxNotReplayableError{}
+}
+
+type cancellationDelayer struct {
+	*sync.Cond
+	delayEnabled bool
+	timer        *time.Timer
+	canceled     bool
+	done         chan struct{}
+}
+
+func newCancellationDelayer() *cancellationDelayer {
+	return &cancellationDelayer{
+		Cond:         sync.NewCond(&sync.Mutex{}),
+		delayEnabled: false,
+		timer:        nil,
+		canceled:     false,
+		done:         make(chan struct{}),
```

Comment:
I was thinking ideally `CleanupCancellationDelayer` would be called at most once. Maybe we can make it closing the channel instead of sending `struct{}` into the channel, to guarantee the function can only be called once. What do you think?